scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSink.java

```diff
@@ -186,29 +256,117 @@ public void startBundle() {
+    static <KeyT> KV<BucketShardId, byte[]> processKey(
+        KeyT key, BucketMetadata<KeyT, ?> metadata, int shardId) {
+      final byte[] keyBytes = metadata.encodeKeyBytes(key);
 
-      final BucketShardId bucketShardId =
-          keyBytes != null
-              ? BucketShardId.of(bucketMetadata.getBucketId(keyBytes), shardId)
-              : BucketShardId.ofNullKey(shardId);
+      return (keyBytes != null)
+          ? KV.of(BucketShardId.of(metadata.getBucketId(keyBytes), shardId), keyBytes)
+          : KV.of(BucketShardId.ofNullKey(shardId), NULL_SORT_KEY);
+    }
 
-      final byte[] sortKey = keyBytes != null ? keyBytes : NULL_SORT_KEY;
+    static <ValueT> byte[] getValueBytes(Coder<ValueT> coder, ValueT value) {
       try {
-        final byte[] valueBytes = CoderUtils.encodeToByteArray(valueCoder, record);
-        c.output(KV.of(bucketShardId, KV.of(sortKey, valueBytes)));
+        return CoderUtils.encodeToByteArray(coder, value);
       } catch (CoderException e) {
-        throw new RuntimeException("Caught coder exception: " + e);
+        throw new RuntimeException(e);
       }
     }
 
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      final V record = c.element();
+      final KV<BucketShardId, byte[]> bucketAndSortKey =
+          processKey(extractKeyFn.apply(record), bucketMetadata, shardId);
+
+      c.output(
+          KV.of(
+              bucketAndSortKey.getKey(),
+              KV.of(bucketAndSortKey.getValue(), getValueBytes(valueCoder, record))));
+    }
+
+    @Override
+    public void populateDisplayData(Builder builder) {
+      super.populateDisplayData(builder);
+      builder.delegate(bucketMetadata);
+    }
+  }
+
+  /** Extract bucket and shard id for grouping, and key bytes for sorting. */
+  private static class ExtractKeysWithCache<K, V>
+      extends DoFnWithResource<
+          V, KV<BucketShardId, KV<byte[], byte[]>>, Cache<K, KV<Integer, byte[]>>> {
+    // Substitute null keys in the output KV<byte[], V> so that they survive serialization
+    private final SerializableFunction<V, K> extractKeyFn;
+    private final Coder<V> valueCoder;
+    private final BucketMetadata<K, ?> bucketMetadata;
+    private final int cacheSize;
+    private transient int shardId;
+    private Counter cacheHits;
+    private Counter cacheMisses;
+
+    ExtractKeysWithCache(
+        BucketMetadata<K, ?> bucketMetadata,
+        SerializableFunction<V, K> extractKeyFn,
+        Coder<V> valueCoder,
+        int cacheSize) {
+      this.bucketMetadata = bucketMetadata;
+      this.extractKeyFn = extractKeyFn;
+      this.valueCoder = valueCoder;
+      this.cacheSize = cacheSize;
+      cacheHits = Metrics.counter(SortedBucketSink.class, "cacheHits");
+      cacheMisses = Metrics.counter(SortedBucketSink.class, "cacheMisses");
+    }
+
+    @Override
+    public ResourceType getResourceType() {
+      return ResourceType.PER_CLASS;
```

Comment:
Should be `PER_INSTANCE` in case there're multiple sinks with the same `K` type?